File: swea/swea_1247.java
// Solve 2024-02-15

// SWEA에 제출할 때는 class 이름을 Solution으로 설정해야 한다.

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class swea_1247 {

    static class Point {
        int x, y;

        Point(int x, int y) {
            this.x = x;
            this.y = y;
        }

        int getDistFrom(Point p) {
            return Math.abs(this.x - p.x) + Math.abs(this.y - p.y);
        }
    }

    static int n;
    static Point[] points;
    static int minSumDist;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;
        int t = Integer.parseInt(br.readLine());

        for (int ti = 1; ti <= t; ti++) {
            n = Integer.parseInt(br.readLine());
            points = new Point[n + 2];
            minSumDist = Integer.MAX_VALUE;
            st = new StringTokenizer(br.readLine(), " ");

            for (int i = 0; i < n + 2; i++) {
                int x = Integer.parseInt(st.nextToken());
                int y = Integer.parseInt(st.nextToken());
                points[i] = new Point(x, y);
            }

            dfs(2, 0, 0, 0);
            sb.append("#").append(ti).append(" ").append(minSumDist).append("\n");
        }

        System.out.print(sb);
        br.close();
    }

    static void dfs(int depth, int prev, int selected, int sumDist) {
        if (sumDist >= minSumDist) return;

        if (depth == n + 2) {
            sumDist += points[1].getDistFrom(points[prev]);

            if (sumDist < minSumDist) {
                minSumDist = sumDist;
            }

            return;
        }

        for (int i = 2; i < n + 2; i++) {
            if ((selected & 1 << i) != 0) continue;

            dfs(depth + 1, i, selected | 1 << i, sumDist + points[i].getDistFrom(points[prev]));
        }
    }

}

File: swea/swea_1247.cpp
// Solve 2024-01-14
// Update 2024-02-15

#include <iostream>
using namespace std;

struct Point{
    int x, y;

    Point(int x = 0, int y = 0) : x(x), y(y) {}

    int get_dist_from(Point &p) {
        return abs(x - p.x) + abs(y - p.y);
    }
};

int n;
Point points[12];
int min_sum_dist;

void dfs(int depth, int prev, int selected, int sum_dist) {
    if (sum_dist >= min_sum_dist) return;

    if (depth == n + 2) {
        sum_dist += points[1].get_dist_from(points[prev]);

        if (sum_dist < min_sum_dist) {
            min_sum_dist = sum_dist;
        }

        return;
    }

    for (int i = 2; i < n + 2; i++) {
        if ((selected & 1 << i) != 0) continue;

        dfs(depth + 1, i, selected | 1 << i, sum_dist + points[i].get_dist_from(points[prev]));
    }
}

int main(int argc, char** argv) {
    int t;
    cin >> t;

    for (int ti = 1; ti <= t; ti++) {
        cin >> n;

        for (int i = 0; i < n + 2; i++) {
            cin >> points[i].x >> points[i].y;
        }

        min_sum_dist = 1000000;
        dfs(2, 0, 0, 0);

        cout << '#' << ti << ' ' << min_sum_dist << '\n';
    }

    return 0;
}
